### PR TITLE
Add Callback for saving pickled job info

### PR DESCRIPTION
**Pull request by jieru-hu**
Add an experimental Callback for pickling job info. Having this as a Callback enables users to opt-in this behavior. 

The doc updates will come after all of #1805 is completed.

### CODE REVIEW

File: hydra/experimental/pikcle_job_info_callback.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class PickleJobInfoCallback(Callback):
+    output_dir: Path
+
+    def on_job_start(self, config: DictConfig, **kwargs: Any) -> None:
+        self.output_dir = Path(config.hydra.runtime.output_dir) / Path(
+            config.hydra.output_subdir
+        )
+        filename = "config.pickle"
+        self._save_pickle(obj=config, filename=filename, output_dir=self.output_dir)
+        log.info(f"Saving job configs in {str(self.output_dir / filename)}")
```
Comment (pixelb) [inline hydra/experimental/pikcle_job_info_callback.py:25]: The explicit str() is redundant I think

File: tests/test_apps/app_with_pickle_job_info_callback/config.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+foo: bar
+
+hydra:
+  callbacks:
+    save_job_info:
+      _target_: hydra.experimental.pikcle_job_info_callback.PickleJobInfoCallback
```
Comment (Jasha10) [inline tests/test_apps/app_with_pickle_job_info_callback/config.yaml:6]:
```suggestion
      _target_: hydra.experimental.pickle_job_info_callback.PickleJobInfoCallback
```

File: hydra/experimental/pikcle_job_info_callback.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    ) -> None:
+        filename = "job_return.pickle"
+        self._save_pickle(obj=job_return, filename=filename, output_dir=self.output_dir)
+        log.info(f"Saving job_return in {str(self.output_dir / filename)}")
+
+    def _save_pickle(self, obj: Any, filename: str, output_dir: Path) -> None:
+        output_dir.mkdir(parents=True, exist_ok=True)
+        assert output_dir is not None
+        with open(str(output_dir / filename), "wb") as file:
+            pickle.dump(obj, file)
```
[the remark this answers is not in the document]
Comment (jieru-hu) [reply hydra/experimental/pikcle_job_info_callback.py]: good point, thanks for the link. reading through the doc it seems like there's some extra benefits in setting it to be the highest, so I update it to that. Let me know what you think  :)

File: hydra/experimental/pikcle_job_info_callback.py (hunk 1 above)
Comment (jieru-hu) [reply hydra/experimental/pikcle_job_info_callback.py]: true that, fixed

File: hydra/experimental/pikcle_job_info_callback.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
```
[the remark this answers is not in the document]
Comment (jieru-hu) [reply hydra/experimental/pikcle_job_info_callback.py]: Did a search internally and didn't find any updates, i think this is good for now

File: hydra/experimental/pickle_job_info_callback.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    ) -> None:
+        filename = "job_return.pickle"
+        self._save_pickle(obj=job_return, filename=filename, output_dir=self.output_dir)
+        log.info(f"Saving job_return in {self.output_dir / filename}")
+
+    def _save_pickle(self, obj: Any, filename: str, output_dir: Path) -> None:
+        output_dir.mkdir(parents=True, exist_ok=True)
+        assert output_dir is not None
+        with open(str(output_dir / filename), "wb") as file:
+            pickle.dump(obj, file, protocol=pickle.HIGHEST_PROTOCOL)
```
Comment (pixelb) [inline hydra/experimental/pickle_job_info_callback.py:38]: Well HIGHEST will break backwards compat between python 3.8 and 3.7.
If you don't envisage maintaining state between different python versions would ever be required cool.
If we do want to maintain compat, then an explicit `4` would be best.

**Review by pixelb — APPROVED** [review]

File: hydra/experimental/pickle_job_info_callback.py (hunk 5 above)
Comment (jieru-hu) [reply hydra/experimental/pickle_job_info_callback.py:38]: i see, that makes sense.

**Pull request merged**